### ISSUE TITLE
Inconsistent handling of abstract types with conjunction of defaultdict

### ISSUE THREAD

**Issue by otaj** [python/mypy#18778]
**Bug Report**

It seems, that when there are multiple abstract types involved in the annotation (`Mapping[Set[int], Sequence[str]]`) while being implemented by `defaultdict(list)` and inferred through followup statements, the inferred value is incompatible (`Incompatible return value type (got "defaultdict[frozenset[int], list[str]]", expected "Mapping[AbstractSet[int], Sequence[str]]")  [return-value]`).

What is more surprising to me, is that if I replace the annotation of `Set` to `int` (i.e. `Mapping[int, Sequence[str]]`) and adapt the code accordingly, `mypy` does not complain about incompatible return values.

I wanted to give `mypy` a hint by annotating the `defaultdict` on creation, however that deprives me of `append` method as `Sequence` does not expose it. 

**To Reproduce**

https://mypy-play.net/?mypy=1.15.0&python=3.11&gist=6225e3e1598f8bc28e836f6f030ae602

**Expected Behavior**

I expect either no error in both above described cases, or error in both cases as I agree the `Sequence` type on the values is a bit abused.

**Actual Behavior**

The case with `Set` as the key annotation fails with incompatible return value type, whereas the simpler annotation does not fail.

**Your Environment**

Python 3.11, mypy 1.15.0 (as recorded in mypy playground link). Surprisingly, this does not happen in `basedmypy`

**Comment by sterliakov** [comment]
It's exactly the same problem as with (occasionally) having to annotate `list` containers. Mapping is invariant in key type (so you aren't getting errors with `int`: there's no ambiguity). Compare with ([playground](https://mypy-play.net/?mypy=1.15.0&python=3.12&gist=b19e716ee553b5b7a0b5c48460158774&flags=strict)) the following:

```python
from collections.abc import Set

def also_fails()-> list[Set[int]]:
    result = list([])
    result.append(frozenset({1,2,3}))
    result.append(frozenset({1,2,3}))
    result.append(frozenset({1,2,3}))
    return result  # E: Incompatible return value type (got "list[frozenset[int]]", expected "list[AbstractSet[int]]")  [return-value]

def passes()-> list[Set[int]]:
    result: list[Set[int]] = list([])
    result.append(frozenset({1,2,3}))
    result.append(frozenset({1,2,3}))
    result.append(frozenset({1,2,3}))
    return result
```
